Update plugins-scenarios.md to Styx 1.0 API (#320)
HotelsDotCom/styx: docs/developer-guide/plugins-scenarios.md

### Synchronously transforming response

This example demonstrates how to synchronously transform a HTTP response. We will

> use a `Eventual.map` method to add an "X-Foo" header to the response.

Reviewer: ```suggestion
call `Eventual.map` to add an "X-Foo" header to the response.
```

}
```

Step 1. We call the `asyncUrlReplacement`, which returns an `Eventual<Url>`.
The `asyncUrlReplacement` wraps a call to the remote service and converts 

> the outcome into a `Eventual`, which is the basis for our response observable.

Author: Removed that last phrase.

Step 2. A call to `pathReplacementService` makes a non-blocking call to the remote key/value store.
Well, at least we pretend to call the key value store, but in this example we'll just return a 
completed future of a constant value.   

Step 3. `CompletableFuture` is converted to an `Eventual`, so that other asynchronous
operations like `chain.proceed` can be bound to it later on.

Step 4. The eventual outcome from the `asyncUrlReplacement` yields a new, modified URL instance.
We'll transform the `request` by substituting the URL path with the new one. 
This is a quick synchronous operation so we'll do it in a `map` operator. 

Step 5. Finally, we will bind the outcome from `chain.proceed` into the response observable.

> Remember that `chain.proceed` returns an `Eventual<LiveHttpResponse>` it is therefore 

Author: 👍

> interface compatible and can be `flatMap`'d to the response observable. The resulting

Author: 👍

}
```

Step 1. We call the `asyncUrlReplacement`, which returns an `Eventual<Url>`.
The `asyncUrlReplacement` wraps a call to the remote service and converts 

> the outcome into a `Eventual`.

Reviewer: ```suggestion
the outcome into an `Eventual`.
```

Step 2. A call to `pathReplacementService` makes a non-blocking call to the remote key/value store.
Well, at least we pretend to call the key value store, but in this example we'll just return a 
completed future of a constant value.   

Step 3. `CompletableFuture` is converted to an `Eventual`, so that other asynchronous
operations like `chain.proceed` can be bound to it later on.

Step 4. The eventual outcome from the `asyncUrlReplacement` yields a new, modified URL instance.
We'll transform the `request` by substituting the URL path with the new one. 
This is a quick synchronous operation so we'll do it in a `map` operator. 

> Step 5. Finally, we will bind the outcome from `chain.proceed` into the response eventual.

Reviewer: I would suggest always using "`Eventual`" to refer to the class, rather than "eventual", as this is a particular concept for Styx, rather than the normal use of the word (which is just an adjective).

Author: ok

}
```

Step 1. We start by calling `chain.proceed(request)` to obtain a response `Eventual`. 

Step 2. A `callTo3rdParty` returns a `CompletableFuture`. It is converted to an `Eventual` with

> `Eventual.from` and bound to response `Eventual`. 

Author: The `flatMap` operator **binds** the returned `Eventual` into the "called" Eventual. 

When the response is available, it is passed to the `flatMap` operator, and the the eventual outcome from the returned `Eventual` then carries on in the chain of operators.

> `Eventual.from` and bound to response `Eventual`. 

Reviewer: OK. Please change
     "bound to response `Eventual` " 
-> "bound to the response `Eventual` "

> `Eventual.from` and bound to response `Eventual`. 

Author: fixed

Streaming HTTP body content can be transformed both synchronously and asynchronously.
However there are some pitfalls you need to know:

>  - Reference counting. Styx exposes the content stream as a `ByteStream` of reference counted

Reviewer: "Reference counting. Styx exposes" -> "Reference counting: Styx exposes"

Perhaps put "Reference counting" in bold?

Author: 👍

>  - **Reference counting.** Styx exposes the content stream as a `ByteStream` of reference counted

Reviewer: > **Reference counting.** Styx exposes the cont

->

> **Reference counting:** Styx exposes the cont

Author: ok

  `Buffer` objects. 
  

>  - Continuity (or discontinuity) of Styx content `ByteStream`. Each content transformation with 

Reviewer: "Continuity (or discontinuity) of Styx content `ByteStream`. Each content" ->
"**Continuity (or discontinuity) of the content `ByteStream`**: each content"

Author: 👍

In this rather contrived example we will transform HTTP response content to 
upper case letters. 

```java
public class MyPlugin extends Plugin {
    @Override
    public Eventual<LiveHttpResponse> intercept(LiveHttpRequest request, Chain chain) {
        return chain.proceed(
                request.newBuilder()
                        .body(byteStream ->                                                              // 1
                                byteStream.map(buf -> {                                                  // 2
                                    String upperCase = new String(buf.content(), UTF_8).toUpperCase();   // 3
                                    return new Buffer(upperCase, UTF_8);
                                }))
                        .build()
        );
    }
}    
```

> A call to `request.newBuilder` opens up a new request builder that allows request to be transformed.

Reviewer: "allows request to" -> "allows the request to"

Author: 👍

The body transformation involves two lambda methods:

  1. A `.body(Function<ByteStream, ByteStream>)` accepts a lambda that modifies the request
     byte stream. Here we provide a lambda that accepts a `ByteStream` and returns another
     by applying a synchronous `map` operator on the stream.
     
  2. The `ByteStream` is modified by applying a `map` operator that synchronously modifies
     each `Buffer` that streams through. The `map` involves another lambda that accepts a
     `Buffer` and returns a modified buffer.
  

>   3. In this example the buffer is converted to an upper case. 

Reviewer: "In this example the buffer is converted to an upper case. " -> "In this example, the content of the buffer is interpreted as a String and converted to upper case. "

Author: 👍

  

### Synchronously transforming streaming response content

This example demonstrates how an HTTP response content can be transformed synchronously.

```java
public class AsyncResponseContentStreamTransformation implements Plugin {
    @Override
    public Eventual<LiveHttpResponse> intercept(LiveHttpRequest request, Chain chain) {
        return chain.proceed(request)
            .map(response ->                                                                        // 1
                response.newBuilder()                         
                    .body(byteStream ->                                                             // 2
                            byteStream.map(buf -> {                                                 // 3
                                String upperCase = new String(buf.content(), UTF_8).toUpperCase();  // 4
                                return new Buffer(upperCase, UTF_8);
                            }))
                    .build());
    }
}
```

> This is quite similar to the request transformation but it involves an additional lambda

Reviewer: "This is quite similar to the request transformation but it involves an additional lambda
expression to capture the HTTP response from `Eventual` envelope:"
->
"This is quite similar to the request transformation, but it involves an additional lambda
expression to capture the HTTP response from its `Eventual` envelope:"

Author: 👍

expression to capture the HTTP response from `Eventual` envelope:

>   1. `chain.proceed` returns an `Eventual` HTTP response. We apply a `Eventual.map` operator

Reviewer: "We apply a `Eventual.map` operator" -> "We apply an `Eventual.map` operator"

Author: 👍

     to tap into this response. The `map` operator accepts a lambda expression that handles the

>      response when it eventually is available.

Reviewer: "when it eventually is available." -> "when it is available."

Author: 👍

> After that, we will transform the response just like we did in the previous example for

Reviewer: Looks like there is some whitespace added between "for" and "response" (newline + spaces?).

Author: Ok. Removed.

In this rather contrived example we will transform HTTP response content to 
upper case letters. 

```java
public class MyPlugin extends Plugin {
    @Override
    public Eventual<LiveHttpResponse> intercept(LiveHttpRequest request, Chain chain) {
        return chain.proceed(
                request.newBuilder()
                        .body(byteStream ->                                                              // 1
                                byteStream.map(buf -> {                                                  // 2
                                    String upperCase = new String(buf.content(), UTF_8).toUpperCase();   // 3
                                    return new Buffer(upperCase, UTF_8);
                                }))
                        .build()
        );
    }
}    
```

A call to `request.newBuilder` opens up a new request builder that allows the request 

> to be transformed to.

Reviewer: It looks like this sentence ended prematurely. Or should it just say:

```suggestion
to be transformed.
```
?

Author: Yes.

The body transformation involves two lambda methods:

>   1. A `.body(Function<ByteStream, ByteStream>)` accepts a lambda that modifies the request

Reviewer: ```suggestion
  1. `.body(Function<ByteStream, ByteStream>)` accepts a lambda that modifies the request
```

     byte stream. Here we provide a lambda that accepts a `ByteStream` and returns another
     by applying a synchronous `map` operator on the stream.
     
  2. The `ByteStream` is modified by applying a `map` operator that synchronously modifies
     each `Buffer` that streams through. The `map` involves another lambda that accepts a
     `Buffer` and returns a modified buffer.
  

>   3. In this example, the content of the buffer buffer is interpreted as `String` and 

Reviewer: > In this example, the content of the buffer buffer is interpreted as `String` and 
     converted to an upper case. 

->

> In this example, the content of the buffer is interpreted as `String` and converted to upper case.

Author: fixed

>   3. In this example, the content of the buffer buffer is interpreted as `String` and 

Reviewer: ```suggestion
  3. In this example, the content of the buffer is interpreted as `String` and 
```

     converted to upper case. 
  

### Synchronously transforming streaming response content

> This example demonstrates how an HTTP response content can be transformed synchronously.

Reviewer: ```suggestion
This example demonstrates how HTTP response content can be transformed synchronously.
```

```java
public class AsyncResponseContentStreamTransformation implements Plugin {
    @Override
    public Eventual<LiveHttpResponse> intercept(LiveHttpRequest request, Chain chain) {
        return chain.proceed(request)
            .map(response ->                                                                        // 1
                response.newBuilder()                         
                    .body(byteStream ->                                                             // 2
                            byteStream.map(buf -> {                                                 // 3
                                String upperCase = new String(buf.content(), UTF_8).toUpperCase();  // 4
                                return new Buffer(upperCase, UTF_8);
                            }))
                    .build());
    }
}
```

> This is quite similar to the request transformation but it involves an additional lambda

Reviewer: ```suggestion
This is quite similar to the request transformation, but it involves an additional lambda
```
